src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionPools.cs

```diff
@@ -0,0 +1,347 @@
+            }
+
+            private async Task<Connection> TryCreateNewConnectionAsync(string serviceName, object callbackTarget, CancellationToken cancellationToken)
+            {
+                var dataRpc = _remotableDataRpc.TryAddReference();
+                if (dataRpc == null)
+                {
+                    // dataRpc is disposed. this can happen if someone killed remote host process while there is
+                    // no other one holding the data connection.
+                    // in those error case, don't crash but return null. this method is TryCreate since caller expects it to return null
+                    // on such error situation.
+                    return null;
+                }
+
+                // get stream from service hub to communicate service specific information
+                // this is what consumer actually use to communicate information
+                var serviceStream = await Connections.RequestServiceAsync(_hubClient, serviceName, _hostGroup, _timeout, cancellationToken).ConfigureAwait(false);
+
+                return new JsonRpcConnection(_hubClient.Logger, callbackTarget, serviceStream, dataRpc);
+            }
+
+            private void Free(string serviceName, JsonRpcConnection connection)
+            {
+                using (_shutdownLock.DisposableRead())
+                {
+                    if (!_usePool)
+                    {
+                        // pool is not being used.
+                        connection.Dispose();
+                        return;
+                    }
+
+                    // queue must exist
+                    var queue = _pools[serviceName];
+                    if (queue.Count >= _maxPoolConnections)
+                    {
+                        // let the connection actually go away
+                        connection.Dispose();
+                        return;
+                    }
+
+                    // pool the connection
+                    queue.Enqueue(connection);
+                }
+            }
+
+            public void Shutdown()
+            {
+                using (_shutdownLock.DisposableWrite())
+                {
+                    // mark not to use pool
+                    _usePool = false;
+
+                    // let ref count this one is holding go
+                    _remotableDataRpc.Dispose();
+
+                    // let all connections in the pool to go away
+                    foreach (var kv in _pools)
+                    {
+                        while (kv.Value.TryDequeue(out var connection))
+                        {
+                            connection.Dispose();
+                        }
+                    }
+                }
+            }
+
+            private class PooledConnection : Connection
```

Comment:
is this guy disposable.  If so, can we have a Dispose method that throws if it is called in a finalizer (i.e. someone forgot to clean it up explicitly).